File: java/SimpleFilters.java
// elprep-bench.
// Copyright (c) 2018-2023 imec vzw.

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version, and Additional Terms
// (see below).

// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public
// License and Additional Terms along with this program. If not, see
// <https://github.com/ExaScience/elprep-bench/blob/master/LICENSE.txt>.

package elprep;

import java.io.*;
import java.lang.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class SimpleFilters {

  private static final Slice SN = new Slice("SN");

  public static Function<SamHeader, Predicate<SamAlignment>> replaceReferenceSequenceDictionary (List<Map<Slice, Slice>> dict) {
    return (header) -> {
      var sortingOrder = header.getHD_SO();
      if (sortingOrder.equals(SamHeader.coordinate)) {
        var previousPos = -1;
        var oldDict = header.SQ;
        for (var entry : dict) {
          var sn = entry.get(SN);
          var pos = SamHeader.find(oldDict, (e) -> sn.equals(e.get(SN)));
          if (pos >= 0) {
            if (pos > previousPos) {
              previousPos = pos;
            } else {
              header.setHD_SO(SamHeader.unknown);
              break;
            }
          }
        }
      }
      var dictTable = new HashSet<Slice>();
      for (var entry : dict) {
        dictTable.add(entry.get(SN));
      }
      header.SQ = dict;
      return (aln) -> dictTable.contains(aln.RNAME);
    };
  }

  public static Function<SamHeader, Predicate<SamAlignment>> replaceReferenceSequenceDictionaryFromSamFile (String samFile) {
    try (var reader = new BufferedReader(new InputStreamReader(new FileInputStream(samFile), "US-ASCII"))) {
      var header = new SamHeader(reader);
      return replaceReferenceSequenceDictionary(header.SQ);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static Predicate<SamAlignment> filterUnmappedReads (SamHeader header) {
    return (aln) -> (aln.FLAG & SamAlignment.Unmapped) == 0;
  }

  private static final Slice star = new Slice("*");

  public static Predicate<SamAlignment> filterUnmappedReadsStrict (SamHeader header) {
    return (aln) -> ((aln.FLAG & SamAlignment.Unmapped) == 0) && (aln.POS != 0) && (!aln.RNAME.equals(star));
  }

  public static Predicate<SamAlignment> filterDuplicateReads (SamHeader header) {
    return (aln) -> (aln.FLAG & SamAlignment.Duplicate) == 0;
  }

  private static final Slice atsr = new Slice("@sr");
  private static final Slice sr = new Slice("sr");

  public static Predicate<SamAlignment> filterOptionalReads (SamHeader header) {
    if (header.userRecords.get(atsr) == null) {
      return null;
    } else {
      header.userRecords.remove(atsr);
      return (aln) -> Fields.assoc(aln.TAGS, sr) == null;
    }
  }

  private static final Slice ID = new Slice("ID");

  public static Function<SamHeader, Predicate<SamAlignment>> addOrReplaceReadGroup (Map<Slice, Slice> readGroup) {
    return (header) -> {
      header.RG = new ArrayList<>();
      header.RG.add(readGroup);
      var id = readGroup.get(ID);
      return (aln) -> {aln.setRG(id); return true;};
    };
  }

  private static final Slice PP = new Slice("PP");

  public static Function<SamHeader, Predicate<SamAlignment>> addPGLine (Map<Slice, Slice> newPG) {
    return (header) -> {
      String[] id = new String[]{newPG.get(ID).toString()};
      while (SamHeader.find(header.PG, (entry) -> id[0].equals(entry.get(ID))) >= 0) {
        id[0] += Integer.toHexString(ThreadLocalRandom.current().nextInt(0x10000));
      }
      newPG.put(ID, new Slice(id[0]));
      for (var pg : header.PG) {
        var nextId = pg.get(ID);
        var pos = SamHeader.find(header.PG, (entry) -> nextId.equals(entry.get(PP)));
        if (pos < 0) {
          newPG.put(PP, nextId); break;
        }
      }
      header.PG.add(newPG);
      return null;
    };
  }

  private static final Slice eq = new Slice("=");

  public static Predicate<SamAlignment> renameChromosomes (SamHeader header) {
    for (var entry : header.SQ) {
      var sn = entry.get(SN);
      if (sn != null) {
        entry.put(SN, new Slice("chr" + sn.toString()));
      }
    }
    return (aln) -> {
      if ((!aln.RNAME.equals(eq)) && (!aln.RNAME.equals(star))) {
        aln.RNAME = new Slice("chr" + aln.RNAME.toString());
      }
      if ((!aln.RNEXT.equals(eq)) && (!aln.RNEXT.equals(star))) {
        aln.RNEXT = new Slice("chr" + aln.RNEXT.toString());
      }
      return true;
    };
  }

  public static Predicate<SamAlignment> addREFID (SamHeader header) {
    var dictTable = new HashMap<Slice, Integer>();
    var index = -1;
    for (var entry : header.SQ) {
      ++index;
      dictTable.put(entry.get(SN), index);
    }
    return (aln) -> {
      var value = dictTable.get(aln.RNAME);
      var val = (value == null ? -1 : value);
      aln.setREFID(val);
      return true;
    };
  }
}

File: cpp/simple_filters.cpp
// elprep-bench.
// Copyright (c) 2018-2023 imec vzw.

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version, and Additional Terms
// (see below).

// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public
// License and Additional Terms along with this program. If not, see
// <https://github.com/ExaScience/elprep-bench/blob/master/LICENSE.txt>.

const string_slice SN("SN");

header_filter replace_reference_sequence_dictionary (const vector<string_map>& dict) {
	return [dict](const shared_ptr<sam_header>& header) -> alignment_filter {
		if (header->get_hd_so() == coordinate) {
			auto previous_pos = -1;
			auto& old_dict = header->sq;
			for (auto& entry: dict) {
				auto it = entry.find(SN);
				if (it == entry.end()) {
					throw runtime_error("SN not found.");
				}
				auto sn = it->second;
				auto pos = find(old_dict, [&sn](const string_map& entry){
						auto it = entry.find(SN);
						return (it != entry.end()) && (it->second == SN);
					});
				if (pos >= 0) {
					if (pos > previous_pos) {
						previous_pos = pos;
					} else {
						header->set_hd_so(unknown);
						break;
					}
				}
			}
		}
		auto dict_table = make_shared<multiset<string_slice>>();
		for (auto& entry: dict) {
			auto it = entry.find(SN);
			if (it == entry.end()) {
				throw runtime_error("SN not found.");
			}
			dict_table->insert(it->second);
		}
		header->sq = dict;
		return [dict_table](const shared_ptr<sam_alignment>& aln) -> bool {
			return dict_table->count(aln->rname) > 0;
		};
	};
}

header_filter replace_reference_sequence_dictionary_from_sam_file (const string& sam_file) {
	ifstream input(sam_file);
	istream_wrapper wrapper(input);
	sam_header header(wrapper);
	return replace_reference_sequence_dictionary(header.sq);
}

alignment_filter filter_unmapped_reads (const shared_ptr<sam_header>&) {
	return [](const shared_ptr<sam_alignment>& aln){
		return (aln->flag & unmapped) == 0;
	};
}

alignment_filter filter_unmapped_reads_strict (const shared_ptr<sam_header>&) {
	return [](const shared_ptr<sam_alignment>& aln){
		return ((aln->flag & unmapped) == 0) && (aln->pos != 0) && (aln->rname != star);
	};
}

alignment_filter filter_duplicate_reads (const shared_ptr<sam_header>&) {
	return [](const shared_ptr<sam_alignment>& aln){
		return (aln->flag & duplicate) == 0;
	};
}

const string_slice sr("sr");
const string_slice at_sr("@sr");

alignment_filter filter_optional_reads (const shared_ptr<sam_header>& header) {
	auto record = header->user_records.find(at_sr);
	if (record == header->user_records.end()) {
		return nullptr;
	} else {
		header->user_records.erase(record);
		return [](const shared_ptr<sam_alignment>& aln){
			return assoc(aln->tags, sr) == aln->tags.end();
		};
	}
}

header_filter add_or_replace_read_group (const string_map& read_group) {
	return [read_group](const shared_ptr<sam_header>& header) -> alignment_filter {
		header->rg = {read_group};
		auto it = read_group.find(ID);
		if (it == read_group.end()) {
			throw runtime_error("ID not found.");
		}
		auto id = it->second;
		return [id](const shared_ptr<sam_alignment>& aln) -> bool {
			aln->set_rg(id); return true;
		};
	};
}

uniform_int_distribution<int> pg_id_dist(0, 0x10000);
random_device rd;

const string_slice PP("PP");

header_filter add_pg_line (string_map& new_pg) {
	return [&new_pg](const shared_ptr<sam_header>& header) -> alignment_filter {
		auto it = new_pg.find(ID);
		if (it == new_pg.end()) {
			throw runtime_error("ID not found.");
		}
		auto id = it->second;
		if (find(header->pg, [&id](const string_map& entry){
					auto it = entry.find(ID);
					return (it != entry.end()) && (it->second == id);
				}) >= 0) {
			ostringstream out; out << id << hex;
			do {
				out << pg_id_dist(rd);
				id = out.str();
			} while (find(header->pg, [&out, &id](const string_map& entry){
						auto it = entry.find(ID);
						return (it != entry.end()) && (it->second == id);
					}) >= 0);
		}
		new_pg.find(ID)->second = id;
		for (auto& pg: header->pg) {
			auto it = pg.find(ID);
			if (it == pg.end()) {
				throw runtime_error("ID not found.");
			}
			auto next_id = it->second;
			auto pos = find(header->pg, [&next_id](const string_map& entry){
					auto it = entry.find(PP);
					return (it != entry.end()) && (it->second == next_id);
				});
			if (pos < 0) {
				new_pg.insert({PP, next_id}); break;
			}
		}
		header->pg.push_back(new_pg);
		return nullptr;
	};
}

alignment_filter add_refid (const shared_ptr<sam_header>& header) {
	auto dict_table = make_shared<unordered_map<string_slice, int32_t>>();
	for (int32_t index = 0; index < header->sq.size(); index++) {
		auto& sq = header->sq[index];
		auto it = sq.find(SN);
		if (it == sq.end()) {
			throw runtime_error("SN not found.");
		}
		dict_table->insert({it->second, index});
	}
	return [dict_table](const shared_ptr<sam_alignment>& aln) -> bool {
		auto it = dict_table->find(aln->rname);
		aln->set_refid((it == dict_table->end()) ? -1 : it->second);
		return true;
	};
}
